pkg/securitycontext/types.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package securitycontext
+
+import (
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+type SecurityContextProvider interface {
+	// ModifyContainerConfig is called before the Docker createContainer call.
+	// The security context provider can make changes to the Config with which
+	// the container is created.
+	// An error is returned if it's not possible to secure the container as
+	// requested with a security context.
+	ModifyContainerConfig(pod *api.Pod, container *api.Container, config *docker.Config) error
+
+	// ModifyHostConfig is called before the Docker runContainer call.
+	// The security context provider can make changes to the HostConfig, affecting
+	// security options, whether the container is privileged, volume binds, etc.
+	// An error is returned if it's not possible to secure the container as requested
+	// with a security context.
+	ModifyHostConfig(pod *api.Pod, container *api.Container, hostConfig *docker.HostConfig)
+}
+
+// SecurityContext specifies the security constraints associated with a service account
+type SecurityContext struct {
```

Comment:
I think so - for step one we wanted to demonstrate the pod side and enforcement.  Step two would be apiserver enforcement.  Step three would be integration with service account.

The mental model I had was like resources - just like there might be an auto sizing resource initializer, there would be an auto enforcing security context initializer (or in the short term, admission controller).  Kubelet might choose to not start a pod until security context was set on each container.

The kubelet policy and the apiserver imposed policy would have strong parallels, perhaps be representable the same way.

> On Apr 1, 2015, at 10:37 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/securitycontext/types.go:
> 
> > +	// The security context provider can make changes to the Config with which
> > +	// the container is created.
> > +	// An error is returned if it's not possible to secure the container as
> > +	// requested with a security context.
> > +	ModifyContainerConfig(pod *api.Pod, container *api.Container, config *docker.Config) error
> > +
> > +	// ModifyHostConfig is called before the Docker runContainer call.
> > +	// The security context provider can make changes to the HostConfig, affecting
> > +	// security options, whether the container is privileged, volume binds, etc.
> > +	// An error is returned if it's not possible to secure the container as requested
> > +	// with a security context.
> > +	ModifyHostConfig(pod *api.Pod, container *api.Container, hostConfig *docker.HostConfig)
> > +}
> > +
> > +// SecurityContext specifies the security constraints associated with a service account
> > +type SecurityContext struct {
> How are you planning to deliver this to Kubelet? It seems like it should work like pods: Kubelet watches apiserver but could also read the config from local files.
> 
> —
> Reply to this email directly or view it on GitHub.
>